generate-aligned-shas.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+/**
+ * A helper script to generate a set of SHAs for aligned runs over time.
+ *
+ * The goal is to generate a comparable, representative set of SHAs that can be
+ * used to compare results from browsers over time.
+ */
+
+const fetch = require('node-fetch');
+const flags = require('flags');
+const fs = require('fs');
+const moment = require('moment');
+
+flags.defineString('from', '2017-08-19', 'Starting date for SHAs');
+flags.defineString('to', moment().format('YYYY-MM-DD'), 'Ending date for SHAs');
+flags.defineString('output', null, 'Output file to write SHAs to. Defaults to {stable, experimental}-shas.txt');
+flags.defineStringList('products', ['chrome', 'safari', 'firefox'], 'Products that must align in the returned SHAs');
+flags.defineBoolean('experimental', false, 'Fetch SHAs for experimental runs rather than stable');
+flags.parse();
+
+const SHAS_API = 'https://wpt.fyi/api/shas?aligned=true';
+
+async function ReadShasFromFile(output) {
+  let shas = new Map();
+
+  // Check whether the file exists; on failure just return the empty map.
+  try {
+    await fs.promises.access(output);
+  } catch (error) {
+    return shas;
+  }
+
+  // Otherwise, read the file and then attempt to parse it as a list of
+  // [date,sha] pairs, newline separated.
+  let data = await fs.promises.readFile(output, 'utf-8');
+  let lines = data.split('\n');
+  for (const line of lines) {
+    if (!line)
+      continue;
+    const parts = line.split(',');
+    shas.set(parts[0], parts[1]);
+  }
+
+  console.log(`Resuing ${shas.size} SHAs from ${output}`);
+  return shas;
+}
+
+async function main() {
+  const experimental = flags.get('experimental');
+
+  let output = flags.get('output');
+  if (!output)
+    output = experimental ? 'experimental-shas.txt' : 'stable-shas.txt';
+
+  let shas = await ReadShasFromFile(output);
+
+  let labels = '&labels=master,';
+  labels += experimental ? 'experimental' : 'stable';
+
+  let products = '';
+  for (const product of flags.get('products')) {
+    products += `&product=${product}`;
+  }
+  const shasUrl = `${SHAS_API}${labels}${products}`;
+  console.log(`Base URL: ${shasUrl}`);
+
+  let from = moment(flags.get('from'));
+  let to = moment(flags.get('to'));
+  console.log(`Fetching SHAs from ${from.format('YYYY-MM-DD')} to ${to.format('YYYY-MM-DD')}`);
+
+  // TODO(smcgruer): This loop is surprisingly slow even when all the SHAs are
+  // cached. I suspect both formatting and incrementing dates may be quite slow.
+  let cachedCount = 0;
+  let before = moment();
+  while (from < to) {
+    const formatted_from = from.format('YYYY-MM-DDT[00:00:00Z]');
+    const formatted_to = from.format('YYYY-MM-DDT[23:59:59Z]');
```

Comment:
https://github.com/web-platform-tests/wpt.fyi/blob/master/api/README.md#apishas doesn't say if `from` and `to` are inclusive to exclusive, but the sensible thing would be for `from` to be inclusive and `to` exclusive, so that we can exactly midnight-to-midnight here and not have to wonder if timestamps include milliseconds that could allow runs from the last second of each day to go missing.

Reply:
Amusingly, looking at the code for wpt.fyi, `/api/shas` is always exclusive `to` but is sometimes inclusive `from` and sometimes exclusive `from`... woo.

So what we should be really afraid of is are we missing runs that ended exactly at midnight? ;).

I don't actually think this is worth worrying about.

Reply:
I've filed https://github.com/web-platform-tests/wpt.fyi/issues/1670. I don't know if any runs would be missed by this, but I've been using "T00:00:00Z" in code like https://github.com/foolip/ecosystem-infra-stats/blob/master/wpt_usage_stats.py and this caught my eye.